### ISSUE TITLE
MySqlCommand.Prepare doesn't work with MySQL 8.0.23 due to query attributes

### ISSUE THREAD

**Issue by andrewb3010** [mysql-net/MySqlConnector#1106]
### Discussed in https://github.com/mysql-net/MySqlConnector/discussions/1105

Hi,

So I'm using .NET core 6.0 and MySqlConnector 2.1.1 (Visual Studio 2022) so the latest of each connecting to MySQL 8.0.

I can create and run commands successfully WITHOUT using .Prepare()

But as soon as I use .Prepare() the command object gets set to a null reference. No exception is thrown other than when trying to subsequently run a method on the command object (which is now null). So the .Prepare() is causing an issue that obviously just wipes the command object, and without reporting an error.

I've tried this in a C# managed project and a C++/CLI project using different variations of commands etc... But all do the same thing. MySQL's log reports that the PREPARE command is being sent to MySQL (successfully) but then the actual MySqlCommand gets set to null for some reason. (Note that the same code but using MySql's own MySql.Data package works find with Prepare but unfortunately it doesn't yet support net core 6.0 and obviously isn't as quick.)

Could anyone help please? Or provide some sample code using .Prepare() that they know to work please?

Thank you in advance.

Here's my code:

Obviously this is C++/CLI but the exact same issue is present when using pure managed C#.

```cpp
MySqlCommand^ cmdTest = gcnew MySqlCommand();
cmdTest->Connection = cn;
cmdTest->CommandType = System::Data::CommandType::Text;

//I've tried using ? and @ptid as the parameter placeholder.
//The command ALWAYS works unless I call Prepare()
cmdTest->CommandText = "select count(0) from dpi_data_manager.master_parts mp "
					+ "where mp.part_type_target_id = ?";

//This line successfully sends a PREPARE statement to MySQL as can be seen in the general log
//I've tried putting this line before and after adding the parameter below
cmdTest->Prepare();

MySqlParameter^ p = cmdTest->Parameters->Add("@ptid", MySqlDbType::Int32);
p->Value = 17;

//This line ALWAYS throws a null reference exception whatever variations I use  - unless I don't call Prepare()
System::Int64 n64 = (System::Int64)cmdTest->ExecuteScalar();
```

**Comment by bgrainger** [comment]
What's the full exception and call stack?

**Comment by andrewb3010** [comment]
> What's the full exception and call stack?

Hi,

Thanks for your quick response.

Actually the command object isn't being set to a null reference - it's just the returned scaler is ALWAYS null when I use .Prepare().

So it was actually raising the exception when I try casting the return to Int64. Now I cast to Int64?

Looking at MySQL's general query log the PREPARE statement IS sent. But the EXECUTE statement never is.

Here's some simple C# code:

MySqlCommand cmd = new("select 1234 where ? = 17", cn) ;
cmd.Prepare();

MySqlParameter p = cmd.Parameters.Add("@ptid", MySqlDbType.Int32);
p.Value = 17;

Int64? i = (Int64?) cmd.ExecuteScalar();

Console.WriteLine(i == null ? "NULL" : i.ToString());

So the console output is always 1234 without the prepare but NULL with the prepare.

Inferring the command and parameter are good (ordinarily).

But whilst Prepare() sends the PREPARE query to MySQL, ExecuteScalar isn't sent to MySQL (it's not in the general query log) IF the command has been prepared beforehand, but IS sent if Prepare() is not used.

Am I missing something in the connection string (I already have IgnorePrepare=false) or other configuration?

Are the parameters more fussy when using Prepare()?

Hope you can help.

Thank you.

**Comment by andrewb3010** [comment]
Just to elimanter parameters from the equation even using this command text exhibits the same behaviour:

MySqlCommand cmd = new("select 1234", cn) ;

With Prepare() returns NULL
Without Prepare() returns 1234

**Comment by bgrainger** [comment]
`git bisect` is indicating that the bug was introduced in https://github.com/mysql-net/MySqlConnector/commit/182ef8e17f526fc7168fcf3cf9c86680730b22c6 (which is new in >= 2.1.0).

This is mentioned in the release notes (https://mysqlconnector.net/overview/version-history/#2-1-0) but the implications aren't spelt out in full. There is a bug in MySQL Server 8.0.23-8.0.25 that breaks clients that use query attributes with prepared statements.

You need to update your server to 8.0.26 or later to fix this bug; is that possible for you?

Otherwise, you have to roll MySqlConnector back to 2.0.0 or avoid `MySqlCommand.Prepare`. (There is currently no flag to opt out of the client-side behaviour; it's dependent on the server advertising it and the client assumes the server implements it properly if it claims support.)

**Comment by andrewb3010** [comment]
Thanks so much for the answer.

(Apologies I didn't click on the query attributes in the release notes.)

Will update MySql later on but for now MySqlConnector 2.0.0 is working fine with Prepare() - thank you.

Just out of interest - hoping to use MySqlConnector 2.1.1 (just to be up to date) - I'm guessing there's no way to disable the server advertising query attributes?

**Comment by bgrainger** [comment]
> I'm guessing there's no way to disable the server advertising query attributes?

I haven't found one.

It may be possible to adjust the protocol for [8.0.23,8.0.25] to avoid sending the data the server doesn't expect; this might resolve the problem with no configuration changes required for MySqlConnector users.

**Comment by andrewb3010** [comment]
That would certainly be very useful.

In my case the upgrades for MySQL are quite involving because whilst they're new releases fix things, they also tend to break also (like this). (Other users might have the same sentiment for MySQL server upgrades.)

So it would be much easier (and more convenient having the far quicker net core compatible MySqlConnector) if there was a quick fix in MySqlConnector?

Just an idea.

But thank you very much for your help in any event.

**Comment by bgrainger** [comment]
Fixed in [2.1.2](https://github.com/mysql-net/MySqlConnector/releases/tag/2.1.2).

**Comment by andrewb3010** [comment]
Thank you very much - works perfectly - and very quick!

Merry Christmas!
